job/common/src/main/java/alluxio/job/meta/JobInfo.java

```diff
@@ -79,52 +78,54 @@ public synchronized int compareTo(JobInfo other) {
    *
    * @param taskId the task id
    */
-  public synchronized void addTask(int taskId) {
-    Preconditions.checkArgument(!mTaskIdToInfo.containsKey(taskId), "");
-    mTaskIdToInfo.put(taskId, new TaskInfo().setJobId(mId).setTaskId(taskId)
-        .setStatus(Status.CREATED).setErrorMessage("").setResult(null));
+  public void addTask(int taskId) {
+    TaskInfo oldValue = mTaskIdToInfo.putIfAbsent(taskId,
+        new TaskInfo().setJobId(mId).setTaskId(taskId).setStatus(Status.CREATED).setErrorMessage("")
+            .setResult(null));
+    Preconditions.checkState(oldValue == null,
```

Comment:
Is this supposed to be `!=`? It will always be `oldValue == null` will always evaluate to `true` if the job doesn't exist yet

Reply:
```suggestion
    Preconditions.checkState(oldValue =! null,
```

Reply:
Also, in the case this does happen, will the caller be able to handle the exception gracefully?

Reply:
Well, we expect it to not exist in the map, so we expect the old value to be null.

Also, there was already a `checkArgument` that basically checked the same condition, so I didn't introduce a new precondition.